### DISCUSSION TITLE
Parent and Subfields

### DISCUSSION THREAD

**Discussion by simplecommerce** [vazco/uniforms#1161] [Q&A] [answered]
Hi,

Quick question about the behavior of a Parent and its subfields.
I am using GraphQL as a bridge.
I have a field called `address` that is defined as an object type with subfields, for e.g.: `streetName`.

In my model, `address` is set to `null`.
I have a field that points to `address.streetName`.

I notice that when I try to type in my field, the modelTransform always returns `null` for the `address` and doesn't seem to convert the structure to 

```
{
   "address": {
      "streetName": ""
   }
}
```

I assume this is normal behavior correct?

I just did a test, if I omit the `address: null` it works correctly.

For what reason would a field being defined as `null` not change to the schema structure when a field or subfield changes?

Sorry if I am unclear.

**Comment by radekmie** [comment, marked as answer]
Hi there! I guess we're talking about `AutoForm` here? If so, the logic we use for changing the model is [here](https://github.com/vazco/uniforms/blob/2cb63a41e589f005df713092bad2864dded13237/packages/uniforms/src/AutoForm.tsx#L65). Here's a quick test to see whether it handles this case or not:

```ts
setIn = (model, key, value) => _.setWith(_.clone(model), key, value, _.clone)
setIn({ address: null }, 'address.streetName', 'Elm') // { address: null }
```

No, it doesn't. I don't think it should either, as `null` is technically not a valid `address`, at least by my standards. Do you think it should handle this case? If so, where is this `null` coming from?

**Reply by simplecommerce** [reply]
Yeah sorry about that I am talking about AutoForm.

Ah ok its the behavior of the code you showed me I understand.

The address value comes from our db, we are adding a new form or modifying an existing one and in some cases if the field doesn't exist or is undefined it might be null.

I was just trying to understand why it wouldn't consider null to be similar to undefined. Because in my mind if its null and you are trying to change a sub field, it shouldn't be null anymore because it needs to respect the structure defined.

Once we understood that this was the case, we simply forced it to be `{}` if its null.

But only wanted to ask to understand the logic behind it.

Do you have a scenario in mind where null would be forced and any other manipulations of subfields should not change it?

Thanks!

**Reply by simplecommerce** [reply]
In your code example `_.setWith(_.clone(model), key, value, _.clone)` what is the purpose of the `_.clone` at the end?
I tested by using `{}` or `Object` and that works if `address` is `null` and I try to apply `address.streetName`, it does create a blank object.

**Reply by radekmie** [reply]
The `clone` at the end ensures that we only shallow clone the objects we have to instead of doing a deep copy of the entire model. Using `Object` there breaks arrays:

```ts
setIn = (model, key, value) => _.setWith(_.clone(model), key, value, _.clone)
setIn({}, 'address.0.streetName', 'Elm') // { address: [{ streetName: "Elm" }] }

setIn = (model, key, value) => _.setWith(_.clone(model), key, value, Object)
setIn({}, 'address.0.streetName', 'Elm') // { address: { 0: { streetName: "Elm" } } }
```
